### DISCUSSION TITLE
DNS reply errors

### DISCUSSION THREAD

**Discussion by koehn** [rspamd/rspamd#5123] [Q&A]
I’m getting errors like this in my log:

```
2024-08-29 20:41:01 #37(controller) <k4k3se>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error’ for bl.score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.bl.score.senderscore.com'(likely DNS spoofing or BL internal issues)
```

I am using a local nameserver instance, as confirmed by `rspamadm configdump`:

```
    dns {
        timeout = 5;
        nameserver [
            "bind:53:10",
        ]
    }
```

(`bind` resolves to a nameserver running bind9 in the same k8s cluster.)

When I query that nameserver for the name mentioned in the log:
```
$ dig 1.0.0.127.bl.score.senderscore.com @bind

; <<>> DiG 9.18.28-1~deb12u2-Debian <<>> 1.0.0.127.bl.score.senderscore.com @bind
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3133
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; COOKIE: 7912797fabab50eb0100000066d0de5c55118a1f18173659 (good)
;; QUESTION SECTION:
;1.0.0.127.bl.score.senderscore.com. IN	A

;; ANSWER SECTION:
1.0.0.127.bl.score.senderscore.com. 10 IN A	127.255.255.255

;; Query time: 23 msec
;; SERVER: 10.43.23.147#53(bind) (UDP)
;; WHEN: Thu Aug 29 20:47:24 UTC 2024
;; MSG SIZE  rcvd: 107
```

It returns the expected response. Does anyone know what’s up?

**Comment by fatalbanana** [comment]
It suggests that the RBL in question blocks your resolver.

**Comment by MichaIng** [comment]
I see the same. But read the message carefully guys:
- The DNS query was successful, rspamd got `NOERROR` and an IP back. So nothing got blocked.
- For `bl.score.senderscore.com` this is of course expected, as it is a valid FQDN.
- For `1.0.0.127.bl.score.senderscore.com`, which is actually checked (the message is sort of confusing, showing the added `1.0.0.127.` sub domain elements only at the end), the resolver however return `127.255.255.255`. In other cases it is `127.0.0.255` or similar, but always a loopback IP. rspamd, however, expects a DNS query error in this case, no answer with loopback IP.
- There is no DNS spoofing or any local hosts override or such in place. You can test with e.g.
  ```sh
  dig '@9.9.9.9' '1.0.0.127.bl.score.senderscore.com'
  ```
  that any public DNS provider returns this IP. So rspamd expects something wrong there.

EDIT: Checking further, it seems that these providers `senderscore.com` etc are expected to return a score for a subdomain appended to their DNS query. But they have rate limits. So large public DNS providers like Quad9, Cloudflare or Google run into these rate limiting, in which case they do not get the expected result. But I am not sure how exactly this is implemented, since everyone just queries DNS root servers in the end, which should return the same IP for any domain and sub domain, isn't it? Got it from here: https://unix.stackexchange.com/a/786801/499600

This info is sort of backed the rspamd docs. But I would have expected this being prominent part of the quick setup guide then, if this is so essentially important 🤔: https://docs.rspamd.com/faq#resolver-setup

EDIT2: I just installed Unbound as recursive DNS resolver, and configured the system as well as rspamd to use it with priority, using `master-slave` method. But I still see the same messages, Unbound returns the same loopback IPs when querying for these hostnames. Not sure, maybe it is just expected but works a little better (I did not count or compare the list of these `'no records with this name' was expected` messages exactly. Probably someone can clarify how large the impact is.

**Comment by vstakhov** [comment]
> `bind` resolves to a nameserver running bind9 in the same k8s cluster.

FYI you could not use anything but IP as nameserver before Rspamd 3.13 (that is still not released). So probably the whole discussion is invalid.

**Reply by MichaIng** [reply]
In my case, at first I did not define any nameserver explicitly, so the host's default `1.1.1.1` was used. Now I use:
```
nameserver = "master-slave:127.0.0.1,1.1.1.1";
```
with `127.0.0.1` being a functional Unbound instance on the same host, used in its `/etc/resolv.conf` as well. Those log messages however remained, until I reduced the log level to warnings.

**Comment by koehn** [comment]
@vstakhov For what it’s worth I still get these errors on 3.14.2, using the same configuration (a locally-hosted, non-recursive nameserver that is only used by Rspamd).

**Comment by hartzell** [comment]
I've been setting up a fresh mail server, and my initial configuration had problems with a variety of list providers, e.g.

```
2026-01-02 23:34:45 #15749(controller) <78ygbp>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for list.dnswl.org while 'no records with this name' was expected when querying for '1.0.0.127.list.dnswl.org'(likely DNS spoofing or BL internal issues)
2026-01-02 23:41:25 #15749(controller) <nxnou4>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dwl.dnswl.org while 'no records with this name' was expected when querying for 'AGQl5k4w.dwl.dnswl.org'(likely DNS spoofing or BL internal issues)
2026-01-02 23:44:29 #15749(controller) <71fyi4>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL intern
```

I updated my DNS configuration to NOT use forwarders and everything cleared up except for senderscore.com.  For them I still see:

```
2026-01-07 09:39:50 #6221(controller) <71fyi4>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL internal issues)
```

They do, indeed, seem to be returning a status of `NOERROR`:

```
dig 1.0.0.127.score.senderscore.com

; <<>> DiG 9.20.17 <<>> 1.0.0.127.score.senderscore.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34991
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;1.0.0.127.score.senderscore.com. IN	A

;; ANSWER SECTION:
1.0.0.127.score.senderscore.com. 10 IN	A	127.0.4.99

;; Query time: 470 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Wed Jan 07 09:55:09 PST 2026
;; MSG SIZE  rcvd: 76
```

As a control, a list provider that was unhappy with my initial config is now happy that I'm not forwarding to shared resolvers:

```
dig 1.0.0.127.list.dnswl.org

; <<>> DiG 9.20.17 <<>> 1.0.0.127.list.dnswl.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 33859
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;1.0.0.127.list.dnswl.org.	IN	A

;; AUTHORITY SECTION:
list.dnswl.org.		2530	IN	SOA	a.ns.dnswl.org. admins.dnswl.org. 1767784714 3600 300 432000 43200

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Wed Jan 07 09:55:52 PST 2026
;; MSG SIZE  rcvd: 101
```

I'm not sure if this is something on senderscore's end or my problem or something in rspamd.

**Reply by hartzell** [reply]
I asked the support folks at validity.com (which seems to own/be senderscore.com), and they replied that it is performing as intended, they use 127.0.0.1 for internal testing and querying that address *should* return something:

> Thank you for providing that! 127.0.0.1 is a Loopback IP and is not a valid IP to [senderscore.org](http://senderscore.org/) because it does not send mail. 
>
> We use these Loopback IPs for testing and monitoring purposes which is why it gives a 127.0.4.99 response. 
>
> I am not familiar with Rspamd and cannot speak to the response there. We recommend that you check with them for more information.

**Reply by koehn** [reply]
FWIW I ended up disabling the monitoring for sender score and it seems to work fine (using a non-recursive DNS resolver):

```
  senderscore {
    enabled = true;
  }

  senderscore_reputation {
    disable_monitoring = true;
  }
```

**Reply by hartzell** [reply]
@koehn -- Am I correct that you just dropped those ~7 lines in a `local.d/rbl.conf` file to override the standard settings?

----

Updated 3:16MST

Looks like putting this in `local.d/rbl.conf` does the trick:

```
rbls {
  senderscore {
    enabled = true;
  }

  senderscore_reputation {
    disable_monitoring = true;
  }
}
```

Using it w/out checking for it seems a bit spooky...

**Reply by hartzell** [reply]
Seems that the functionality is documented here: https://docs.rspamd.com/modules/rbl/#monitoring

**Reply by vstakhov** [reply]
Maybe you can write those guys that there is such a thing in the Internet, you know, called standards: https://www.rfc-editor.org/rfc/rfc5782.html#section-5:

```
[5](https://www.rfc-editor.org/rfc/rfc5782.html#section-5).  Test and Contact Addresses

   IPv4-based DNSxLs MUST contain an entry for 127.0.0.2 for testing
   purposes.  IPv4-based DNSxLs MUST NOT contain an entry for 127.0.0.1.

   DNSBLs that return multiple values SHOULD have multiple test
   addresses so that, for example, a DNSBL that can return 127.0.0.5
   would have a test record for 127.0.0.5 that returns an A record with
   the value 127.0.0.5, and a corresponding TXT record.

   IPv6-based DNSxLs MUST contain an entry for ::FFFF:7F00:2 (::FFFF:
   127.0.0.2), and MUST NOT contain an entry for ::FFFF:7F00:1 (::FFFF:
   127.0.0.1), the IPv4-Mapped IPv6 Address [[RFC4291](https://www.rfc-editor.org/rfc/rfc4291)] equivalents of the
   IPv4 test addresses.
```

If they list 127.0.0.1 it's THEIR problem, as they clearly violate RFC. I might just remove their RBL from the default configuration for being broken.

**Reply by hartzell** [reply]
I followed up in the support ticket, pointed them to this thread.

**Reply by vstakhov** [reply]
The reasoning behind this are very simple - we need to ensure that some list is still alive and does not, for example, blacklist the whole Internet. These cases have happened several times in the past. That's why Rspamd has this quite a complicated system of live monitoring. And every DNS list that violates standards creates a potential issue in future.

**Reply by MichaIng** [reply]
I just lowered the log level again to check back our end, and `senderscore.com` is not the only one giving unexpected response:
```console
user@host:/etc/rspamd# journalctl -u rspamd --grep rspamd_monitored_dns_cb
Jan 08 11:23:00 host rspamd[1578028]: <s4ti79>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL internal issues)
Jan 08 11:23:00 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:23:33 host rspamd[1578028]: <s4ti79>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL internal issues)
Jan 08 11:23:46 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:24:01 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for '13Aui2tF.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:25:09 host rspamd[1578028]: <s4ti79>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL internal issues)
Jan 08 11:25:29 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for 'WUbWsq1U.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:25:41 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:26:16 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for 'quS3bhkK4d6n7p.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:28:03 host rspamd[1578028]: <s4ti79>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL internal issues)
Jan 08 11:28:57 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:29:21 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for 'RuhTfvxs_.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:33:11 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:34:30 host rspamd[1578028]: <s4ti79>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for score.senderscore.com while 'no records with this name' was expected when querying for '1.0.0.127.score.senderscore.com'(likely DNS spoofing or BL internal issues)
Jan 08 11:37:10 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for 'MXz8qnmWIs1a6q6.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
user@host:/etc/rspamd# dig 1.0.0.127.zen.spamhaus.org

; <<>> DiG 9.20.15-2-Debian <<>> 1.0.0.127.zen.spamhaus.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 31579
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;1.0.0.127.zen.spamhaus.org.    IN      A

;; ANSWER SECTION:
1.0.0.127.zen.spamhaus.org. 1349 IN     A       127.255.255.254

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Thu Jan 08 11:38:12 CET 2026
;; MSG SIZE  rcvd: 71

user@host:/etc/rspamd# dig 13Aui2tF.dbl.spamhaus.org

; <<>> DiG 9.20.15-2-Debian <<>> 13Aui2tF.dbl.spamhaus.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58146
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;13Aui2tF.dbl.spamhaus.org.     IN      A

;; ANSWER SECTION:
13Aui2tF.dbl.spamhaus.org. 1242 IN      A       127.255.255.254

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Thu Jan 08 11:38:19 CET 2026
;; MSG SIZE  rcvd: 70
```
The used DNS resolver `127.0.0.1#53` is a local recursive Unbound instance. Our IP is our own globally routable IP, nothing shared or so.

Checking for other logs about `spamhaus.org`:
```console
user@host:/etc/rspamd# journalctl -u rspamd --grep spamhaus.org
Jan 08 11:25:41 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_propagate_error: invalid return on resolving zen.spamhaus.org, disable object
Jan 08 11:26:16 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for 'quS3bhkK4d6n7p.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:26:16 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_propagate_error: invalid return on resolving dbl.spamhaus.org, disable object
Jan 08 11:28:57 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:29:21 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for dbl.spamhaus.org while 'no records with this name' was expected when querying for 'RuhTfvxs_.dbl.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:33:11 host rspamd[1578028]: <c9hxbz>; monitored; rspamd_monitored_dns_cb: DNS reply returned 'no error' for zen.spamhaus.org while 'no records with this name' was expected when querying for '1.0.0.127.zen.spamhaus.org'(likely DNS spoofing or BL internal issues)
Jan 08 11:35:28 host rspamd[1578028]: <1jqufq>; monitored; rspamd_monitored_propagate_success: restoring dbl.spamhaus.org after 552.5 seconds of downtime, total downtime: 552.5
```
Hence at some point, the two lists are disabled. Not sure whether it is expected or reasonable to monitor disabled lists, and then enable them again despite giving unexpected replies? Anyway @hartzell `spamhaus` is enabled in your case and gives expected replies? It is the first entry in unmodified `modules.d/rbl.conf` in our case, with rspamd v3.13.2.

**Reply by koehn** [reply]
@hartzell you can see my whole configuration here: https://git.koehn.com/k3s-home/mail/src/branch/main/rspamd

**Reply by hartzell** [reply]
@koehn -- Thank you for the pointer!

**Comment by japc** [comment]
@MichaIng Maybe a glitch? Because I tested those records and all return `NXDOMAIN`, as expected:

```sh
$ dig +noall +comments +answer 1.0.0.127.zen.spamhaus.org @a.gns.spamhaus.org
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 40914
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 2048
```

Either that or some problem with the resolver (I did the test directly to the domain authoritative).

**Comment by japc** [comment]
It's related to the query source: https://www.spamhaus.org/resource-hub/dnsbl/using-our-public-mirrors-check-your-return-codes-now/

**Reply by MichaIng** [reply]
So my answer suggests I am using a public resolver, even when querying `a.gns.spamhaus.org` directly:
```console
$ dig +noall +comments +answer 1.0.0.127.zen.spamhaus.org @a.gns.spamhaus.org
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 65397
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 2048
;; ANSWER SECTION:
1.0.0.127.zen.spamhaus.org. 2100 IN     A       127.255.255.254
```
How can this be? 🤔

EDIT: <s>"or an IP address with generic, unattributable reverse DNS": Our IP returns the valid domain of our mail/MX server domain to PTR requests, otherwise it wouldn't work at all. So that shouldn't be an issue either.</s> EDIT2: It is about the IP of the DNS resolver/server, which should be `a.gns.spamhaus.org` with above command, isn't it?

**Reply by MichaIng** [reply]
Oh, it is IPv4 vs IPv6 in my case. Queries via IPv6 return the unexpected A record, queries via IPv4 the expected NXDOMAIN:
```console
user@host:/etc/apt# dig -4 +noall +comments +answer 1.0.0.127.zen.spamhaus.org @a.gns.spamhaus.org
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 5407
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 2048
user@host:/etc/apt# dig -4 +noall +comments +answer 1.0.0.127.zen.spamhaus.org
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 23167
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
```
I guess this is unintended, or maybe they do not check full IPv6 addresses but (partial) prefixes, and rate those from larger VPS providers as public prefixes or so? @japc since you do not define `-4` explicitly, has your system a public IPv6 address (in which case IPv6 queries should be the default) or prefers IPv4 somehow? I'll test enforcing IPv4 in our Unbound for DNS requests.

EDIT: I just verified that our public IPv4 and IPv6 addresses both return the correct and same PTR hostname of our mailserver.

**Reply by japc** [reply]
> EDIT2: It is about the IP of the DNS resolver/server, which should be `a.gns.spamhaus.org` with above command, isn't it?

No, it should be about the IP from which they see the query coming, that is, the IP of the resolver or host where you did the dig (if directly to gns.spamhaus.org).

To me it returns NXDOMAIN either querying IPv4 DNS or IPv4. But, again, can be related to the PTR on IPv4 vs IPv6.

**Reply by MichaIng** [reply]
Okay, so if Unbound acts as recursive resolver using its internal root hints (matching https://www.internic.net/domain/named.root, i.e. querying DNS root servers), does the "recursive" resolving imply that it queries `a.gns.spamhaus.org` as this domain's authoritative name server? Sorry for the questions, not related to rspamd anymore. I guess I need to ask Spamhaus then why our (unique) IPv6 address with valid PTR is treated as public resolver their end.

And indeed `do-ip6: no` or `prefer-ipv4: yes` in Unbound solves the issue our end, just to verify that the issue really is our IPv6 address.

EDIT: I sent Spamhaus a message about it now, assuming above is the correct interpretation of the issue. We moved our server ~half a year ago, to a quite known VPS provider. So it is thinkable that this IP or its subnet was (mis)used as open resolver before.

**Reply by MichaIng** [reply]
I got reply from Spamhaus, and indeed some larger Hetzner IPv6 address space seems to be blocked or throttled at least (though in our case is appears as permanent), due to high volumes of abusive queries seem from that VPS Provider's network.

Aside of doing queries via IPv4, another alternative is registering for the Spamhaus DQS service, which is for free for low query volumes (below 100,000 per day). One gets a unique key for unique DBL domains, where need to be applied to rspamd settings: https://docs.spamhaus.com/datasets/docs/source/70-access-methods/data-query-service/015-migrating-to-dqs.html

I will test that and post the respective `local.d/rbl.conf` for reference, since I am surely not the only one running an email server at a Hetzner VPS/server.

EDIT: Spamhaus solved the issue for the Hetzner IPv6 range. Hence also with the public DBL, queries via IPv6 do now return the expected response.

**Comment by MichaIng** [comment]
rspamd v4.0.0 seems to have addressed the `score.senderscore.com`, disabling it by default with some meaningful info:
```
    senderscore_reputation {
      # Disabled by default. Requires a registered MyValidity account to function:
      # unregistered IPs receive 127.255.255.255 (blocked) regardless of query volume.
      # See https://knowledge.validity.com/hc/en-us/articles/20961730681243.
      enabled = false;
      symbol = "RBL_SENDERSCORE_REPUT_UNKNOWN";
      checks = ["from"];
      rbl = "score.senderscore.com";
      returncodes_matcher = "luapattern";
      returncodes {
        RBL_SENDERSCORE_REPUT_0 = "127%.0%.4%.%d";
        RBL_SENDERSCORE_REPUT_1 = "127%.0%.4%.1%d";
        RBL_SENDERSCORE_REPUT_2 = "127%.0%.4%.2%d";
        RBL_SENDERSCORE_REPUT_3 = "127%.0%.4%.3%d";
        RBL_SENDERSCORE_REPUT_4 = "127%.0%.4%.4%d";
        RBL_SENDERSCORE_REPUT_5 = "127%.0%.4%.5%d";
        RBL_SENDERSCORE_REPUT_6 = "127%.0%.4%.6%d";
        RBL_SENDERSCORE_REPUT_7 = "127%.0%.4%.7%d";
        RBL_SENDERSCORE_REPUT_8 = "127%.0%.4%.8%d";  # Neutral reputation (80-89).
        RBL_SENDERSCORE_REPUT_9 = ["127%.0%.4%.9%d", "127%.0%.4%.100"];  # Good reputation (90-100).
        RBL_SENDERSCORE_REPUT_BLOCKED = "127%.255%.255%.255";
      }
    }
```
